<?xml version="1.0"?>
<robot name="robot_car" xmlns:xacro="http://www.ros.org/wiki/xacro">
    <xacro:property name="box_x" value="2.2"/>
    <xacro:property name="box_y" value="1.4"/>
    <xacro:property name="box_z" value="0.3"/>
    <xacro:property name="box_M" value="1.1"/>
    <xacro:property name="wheel_steer_r" value="0.15"/>
    <xacro:property name="wheel_steer_l" value="0.05"/>
    <xacro:property name="wheel_r" value="0.35"/>
    <xacro:property name="wheel_l" value="0.2"/>
    <xacro:property name="wheel_M" value="0.8"/>
    <xacro:property name="sphere_r" value="0.25"/>    
    <xacro:property name="sphere_M" value="1.0"/>    
    <xacro:property name="I_sphere" value="${(2.0/5.0)*sphere_M*sphere_r*sphere_r}"/>

    <xacro:property name="ixx" value="${(1/12.0)*box_M*(box_y*box_y + box_z*box_z)}"/>
    <xacro:property name="iyy" value="${(1/12.0)*box_M*(box_x*box_x + box_z*box_z)}"/>
    <xacro:property name="izz" value="${(1/12.0)*box_M*(box_x*box_x + box_y*box_y)}"/>



    <link name="base_link">
        <visual>
            <geometry>
                <box size="${box_x} ${box_y} ${box_z}"/>
            </geometry>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <material name="gray">
                <color rgba="0.5 0.5 0.5 1"/>
            </material>
        </visual>
        <collision>
            <geometry>
                <box size="1.9 ${box_y} ${box_z}"/>
            </geometry>
            <origin xyz="0 0 0" rpy="0 0 0"/>
              <surface>
                <friction>
                    <ode>
                        <mu>1.0</mu>
                        <mu2>1.0</mu2>
                    </ode>
                </friction>
            </surface>
        </collision>
        <inertial>
            <mass value="${box_M}"/>
            <inertia ixx="${ixx}" iyy="${iyy}" izz="${izz}" ixy="0.0" ixz="0.0" iyz="0.0"/>
            <origin xyz="0 0 0" rpy="0 0 0"/>
        </inertial>


    </link>

     <link name="left_wheel_steer">
        <visual>
            <geometry>
                <cylinder radius="${wheel_steer_r}" length="${wheel_steer_l}"/>
            </geometry>
            <origin xyz="0 0 0" rpy="1.5708 0 0"/>
            <material name="gray">
                <color rgba="0.5 0.5 0.5 1"/>
            </material>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="${wheel_steer_r}" length="${wheel_steer_l}"/>
            </geometry>
            <surface>
                <friction>
                    <ode>
                        <mu>1</mu>
                        <mu2>1</mu2>
                    </ode>
                </friction>
            </surface>
            <origin xyz="0 0 0" rpy="1.5708 0 0"/>
        </collision>
        <inertial>
            <mass value="${wheel_M}"/>
            <inertia 
            ixx="${(1.0/12.0)*wheel_M*(3*wheel_steer_r*wheel_steer_r + wheel_steer_l*wheel_steer_l)}"
            iyy="${(1.0/12.0)*wheel_M*(3*wheel_steer_r*wheel_steer_r + wheel_steer_l*wheel_steer_l)}"
            izz="${0.5*wheel_M*wheel_steer_r*wheel_steer_r}"
            ixy="0" ixz="0" iyz="0"/>
            <origin xyz="0 0 0" rpy="0 0 0"/>
        </inertial>
    </link>

    <link name="right_wheel_steer">
        <visual>
            <geometry>
                <cylinder radius="${wheel_steer_r}" length="${wheel_steer_l}"/>
            </geometry>
            <origin xyz="0 0 0" rpy="1.5708 0 0"/>
            <material name="gray">
                <color rgba="0.5 0.5 0.5 1"/>
            </material>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="${wheel_steer_r}" length="${wheel_steer_l}"/>
            </geometry>
            <surface>
                <friction>
                    <ode>
                        <mu>1</mu>
                        <mu2>1</mu2>
                    </ode>
                </friction>
            </surface>
            <origin xyz="0 0 0" rpy="1.5708 0 0"/>
        </collision>
        <inertial>
            <mass value="${wheel_M}"/>
            <inertia 
            ixx="${(1.0/12.0)*wheel_M*(3*wheel_steer_r*wheel_steer_r + wheel_steer_l*wheel_steer_l)}"
            iyy="${(1.0/12.0)*wheel_M*(3*wheel_steer_r*wheel_steer_r + wheel_steer_l*wheel_steer_l)}"
            izz="${0.5*wheel_M*wheel_steer_r*wheel_steer_r}"
            ixy="0" ixz="0" iyz="0"/>
            <origin xyz="0 0 0" rpy="0 0 0"/>
        </inertial>
    </link>


    <link name="left_wheel">
        <visual>
            <geometry>
                <cylinder radius="${wheel_r}" length="${wheel_l}"/>
            </geometry>
            <origin xyz="0 0 0" rpy="1.5708 0 0"/>
            <material name="gray">
                <color rgba="0.5 0.5 0.5 1"/>
            </material>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="${wheel_r}" length="${wheel_l}"/>
            </geometry>
            <surface>
                <friction>
                    <ode>
                        <mu>1</mu>
                        <mu2>1</mu2>
                    </ode>
                </friction>
            </surface>
            <origin xyz="0 0 0" rpy="1.5708 0 0"/>
        </collision>
        <inertial>
            <mass value="${wheel_M}"/>
            <inertia 
            ixx="${(1.0/12.0)*wheel_M*(3*wheel_r*wheel_r + wheel_l*wheel_l)}"
            iyy="${(1.0/12.0)*wheel_M*(3*wheel_r*wheel_r + wheel_l*wheel_l)}"
            izz="${0.5*wheel_M*wheel_r*wheel_r}"
            ixy="0" ixz="0" iyz="0"/>
            <origin xyz="0 0 0" rpy="0 0 0"/>
        </inertial>
    </link>

    <link name="right_wheel">
        <visual>
            <geometry>
                <cylinder radius="${wheel_r}" length="${wheel_l}"/>
            </geometry>
            <origin xyz="0 0 0" rpy="1.5708 0 0"/>
            <material name="gray">
                <color rgba="0.5 0.5 0.5 1"/>
            </material>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="${wheel_r}" length="${wheel_l}"/>
            </geometry>
            <surface>
                <friction>
                    <ode>
                        <mu>1</mu>
                        <mu2>1</mu2>
                    </ode>
                </friction>
            </surface>
            <origin xyz="0 0 0" rpy="1.5708 0 0"/>
        </collision>
        <inertial>
            <mass value="${wheel_M}"/>
            <inertia 
            ixx="${(1.0/12.0)*wheel_M*(3*wheel_r*wheel_r + wheel_l*wheel_l)}"
            iyy="${(1.0/12.0)*wheel_M*(3*wheel_r*wheel_r + wheel_l*wheel_l)}"
            izz="${0.5*wheel_M*wheel_r*wheel_r}"
            ixy="0" ixz="0" iyz="0"/>
            <origin xyz="0 0 0" rpy="0 0 0"/>
        </inertial>
    </link>

    <link name="sphere">
        <visual>
            <geometry>
                <sphere radius="${sphere_r}"/>
            </geometry>
            <origin xyz="0 0 0" rpy="1.5708 0 0"/>
            <material name="gray">
                <color rgba="0.5 0.5 0.5 1"/>
            </material>
        </visual>
        <collision>
            <geometry>
                <sphere radius="${sphere_r}"/>
            </geometry>
            <surface>
                <friction>
                    <ode>
                        <mu>1.0</mu>
                        <mu2>1.0</mu2>
                    </ode>
                </friction>
            </surface>
            <origin xyz="0 0 0" rpy="1.5708 0 0"/>
        </collision>
        <inertial>
        <mass value="${sphere_M}"/>
        <inertia 
            ixx="${I_sphere}" 
            iyy="${I_sphere}" 
            izz="${I_sphere}" 
            ixy="0" ixz="0" iyz="0"/>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        </inertial>


    </link>

    <!-- Depth camera link -->
    <link name="depth_camera_link">
    <visual>
        <geometry>
        <box size="0.05 0.05 0.05"/>  <!-- camera size -->
        </geometry>
        <material name="black">
        <color rgba="0.0 0.0 0.0 1.0"/>
        </material>
    </visual>
    <inertial>
        <mass value="0.1"/>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <inertia ixx="1e-5" iyy="1e-5" izz="1e-5" ixy="0" ixz="0" iyz="0"/>
    </inertial>
    </link>



    <joint name="left_body_steer" type="revolute">
        <parent link="base_link"/>
        <child link="left_wheel_steer"/>
        <origin xyz="0.5 0.8 0.0" rpy="0 0 0"/>
        <axis xyz="0 0 1"/>
        <limit lower="-1.57" upper="1.57" effort="1" velocity="1"/>
    </joint>

    <joint name="right_body_steer" type="revolute">
        <parent link="base_link"/>
        <child link="right_wheel_steer"/>
        <origin xyz="0.5 -0.8 0.0" rpy="0 0 0"/>
        <axis xyz="0 0 1"/>
        <limit lower="-1.57" upper="1.57" effort="1" velocity="1"/>
    </joint>


    <joint name="left_body_wheel" type="continuous">
        <parent link="left_wheel_steer"/>
        <child link="left_wheel"/>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <axis xyz="0 1 0"/>
        <limit effort="1" velocity="5"/>
    </joint>

    <joint name="right_body_wheel" type="continuous">
        <parent link="right_wheel_steer"/>
        <child link="right_wheel"/>
        <origin xyz="0.0 0.0 0.0" rpy="0 0 0"/>
        <axis xyz="0 1 0"/>
        <limit effort="1" velocity="5"/>
    </joint>

    <joint name="body_sphere" type="continuous">
        <parent link="base_link"/>
        <child link="sphere"/>
        <origin xyz="-0.7 0 -0.15" rpy="0 0 0"/>
        <axis xyz="0 1 0"/>
        <limit effort="10" velocity="10"/>
    </joint>

    <!-- Fixed joint attaching the camera to the robot -->
    <joint name="depth_camera_joint" type="fixed">
        <parent link="base_link"/>  <!-- or wherever you want to mount it -->
        <child link="depth_camera_link"/>
        <origin xyz="1.0 0 0.3" rpy="0 0 0"/> <!-- Adjust position -->
    </joint>



    <ros2_control name="GazeboSystem" type="system">
        <hardware>
            <plugin>gazebo_ros2_control/GazeboSystem</plugin>
        </hardware>

        <joint name="left_body_steer">
            <command_interface name="position"/>
            <state_interface name="position"/>
        </joint>

        <joint name="right_body_steer">
            <command_interface name="position"/>
            <state_interface name="position"/>
        </joint>

        <joint name="body_sphere">
            <command_interface name="velocity"/>
            <state_interface name="velocity"/>
        </joint>
    </ros2_control>

    <gazebo>
        <plugin name="gazebo_ros2_control" filename="libgazebo_ros2_control.so">
            <parameters>/home/song/ros2_ws/src/follower/ros2_controller.yaml</parameters>
        </plugin>
    </gazebo>

    <gazebo reference="depth_camera_link">
        <sensor name="depth_camera" type="depth">
            <always_on>true</always_on>
            <update_rate>30.0</update_rate>
            <visualize>true</visualize>
            <camera>
            <horizontal_fov>1.047</horizontal_fov>  <!-- ~60 degrees -->
                <image>
                    <width>640</width>
                    <height>480</height>
                    <format>R8G8B8</format>
                </image>
                <clip>
                    <near>0.1</near>
                    <far>10.0</far>
                </clip>
            </camera>
            <plugin name="gazebo_ros_depth_camera" filename="libgazebo_ros_camera.so">
                <ros>
                    <namespace>/</namespace>
                </ros>
                <camera_name>depth_camera</camera_name>
                <frame_name>depth_camera_link</frame_name>
                <image_topic_name>image_raw</image_topic_name>
                <depth_image_topic_name>depth/image_raw</depth_image_topic_name>
                <camera_info_topic_name>camera_info</camera_info_topic_name>
            </plugin>
        </sensor>
    </gazebo>



</robot>